quickwit/quickwit-search/src/find_trace_ids_collector.rs

```diff
@@ -0,0 +1,508 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::cmp::{Ord, Ordering};
+use std::collections::{BinaryHeap, HashSet};
+use std::hash::{Hash, Hasher};
+use std::sync::Arc;
+
+use base64::prelude::{Engine, BASE64_STANDARD};
+use base64::DecodeSliceError;
+use fnv::{FnvHashMap, FnvHashSet};
+use itertools::Itertools;
+use quickwit_common::binary_heap;
+use serde::{Deserialize, Serialize};
+use tantivy::collector::{Collector, SegmentCollector};
+use tantivy::fastfield::{Column, MultiValuedFastFieldReader};
+use tantivy::{DateTime, DocId, InvertedIndexReader, Score, SegmentReader};
+
+type B64TraceId = [u8; 24];
+type TermOrd = u64;
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+#[serde(into = "TraceIdFruitForSerde")]
+#[serde(try_from = "TraceIdFruitForSerde")]
+pub struct TraceIdFruit {
+    pub trace_id: B64TraceId,
+    pub span_timestamp: DateTime,
+}
+
+impl TraceIdFruit {
+    pub fn new(trace_id_bytes: &[u8], span_timestamp: DateTime) -> Self {
+        let mut trace_id: B64TraceId = Default::default();
+        trace_id.copy_from_slice(trace_id_bytes);
+
+        Self {
+            trace_id,
+            span_timestamp,
+        }
+    }
+
+    #[cfg(any(test, feature = "testsuite"))]
+    fn for_test(trace_id_bytes: &[u8], span_timestamp_micros: i64) -> Self {
+        let span_timestamp = DateTime::from_timestamp_micros(span_timestamp_micros);
+        Self::new(trace_id_bytes, span_timestamp)
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+struct TraceIdFruitForSerde {
+    trace_id: String,
+    span_timestamp: i64,
+}
+
+impl From<TraceIdFruit> for TraceIdFruitForSerde {
+    fn from(trace_id_fruit: TraceIdFruit) -> Self {
+        Self {
+            trace_id: BASE64_STANDARD.encode(&trace_id_fruit.trace_id),
+            span_timestamp: trace_id_fruit.span_timestamp.into_timestamp_micros(),
+        }
+    }
+}
+
+impl TryFrom<TraceIdFruitForSerde> for TraceIdFruit {
+    type Error = DecodeSliceError;
+
+    fn try_from(for_serde: TraceIdFruitForSerde) -> Result<Self, Self::Error> {
+        let mut trace_id: B64TraceId = Default::default();
+        let num_bytes = BASE64_STANDARD.decode_slice(for_serde.trace_id, &mut trace_id)?;
+        debug_assert_eq!(num_bytes, 24);
+
+        Ok(Self {
+            trace_id,
+            span_timestamp: DateTime::from_timestamp_micros(for_serde.span_timestamp),
+        })
+    }
+}
+
+impl Hash for TraceIdFruit {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.trace_id.hash(state);
+    }
+}
+
+impl Ord for TraceIdFruit {
+    fn cmp(&self, other: &Self) -> Ordering {
+        self.span_timestamp
+            .cmp(&other.span_timestamp)
```

Comment:
I'm not super comfortable with having Ord inconsistent with Eq.
Is it necessary?